### PR TITLE
feat[compositions]: realtime compositor – part 2: changes to MRs

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

Follow-up of https://github.com/crossplane/crossplane/pull/4582: also reconcile XRs when MRs change. This PR starts informers dynamically for GVRs used in XRs, including function XRs.

This feature is under the `EnableRealtimeCompositions` feature gate (`--enable-realtime-compositions`).

Design:
```
                    ┌─────────────────┐
                    │                 │
                    │     Manager     │
                    │                 │
                    │    ┌───────┐    │
                    │    │       │    │
                    └────┤ Cache ├────┘
                         │       │
                         └───▲───┘◄─────┬──────────┬───────────────────┐
                             │          │          │                   │
                    ┌────────┴────────┐ │ ┌────────┴────────┐ ┌────────┴────────┐
       ┌────────────┤ APIExtensions   │ │ │ APIExtensions   │ │ APIExtensions   │
       │ Composed   │ Definition      │ │ │ Compositions    │ │ Offered         │
┌──────┤ Resource   │ Controller      │ │ │ Controller      │ │ Controller      │ ...
│      │ Informers  │                 │ │ │                 │ │                 │
│      └─────┬──────┤  ┌───────────┐  │ │ │                 │ │                 │
│    creates │      └──┤           ├──┘ │ └─────────────────┘ └─────────────────┘
│  & deletes │         │ GVK Routed│    │
│     ┌──────▼────┐    │ Cache     ├────┘
│     │ Composed  │◄───┤           │fallback
│     │ Resource  │ N:1└─────▲─────┘
│     │ Cache     │          │
│     └───────────┘          │
│                            │ N:1 (dynamically started when an XRD changes)
│                   ┌────────┴────────┐
│                   │ APIExctensions  │
│ informs about gvks│ Composite       │
└───────────────────┤ Controller      │
                    │                 │
                    │                 │
                    └─────────────────┘
```

tl/dr: 
1. the APIExtensions definitions controller (which spawns the composite controllers dynamically through the controller engine), maintains a dynamic set of cached (= informers) per gvk. 
2. The composite controllers report (via the `KindObserver` interface) to the definition controller which GVKs the referenced objects have. For those, the "composedResourceInformers" (part of the definition controller) will start informers as part of a cache. 
3. When those caches become synced, they are registered in the GVK-routed cache as delegates. From that moment on that cache serves those objects. The composite controllers use that GVK-routed-cache for their lookups.
4. Last but not least the "composedResourceInformers" distributes events of changed composed resources to the right composite controllers. For that, an index is added to the XR informer by GVK of the referenced object of each XR instance. Through that it is efficient to find the right XRs for every event.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~~Added `backport release-x.y` labels to auto-backport this PR, if necessary.~~
- [ ] ~~Opened a PR updating the [docs], if necessary.~~

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

**Comment by negz** [comment]
Adding a reference to https://github.com/crossplane/crossplane/issues/4316, which I think is related.

File: cmd/crossplane/core/core.go [hunk 1]
```diff
@@ -108,6 +108,7 @@ type startCommand struct {
 	EnableExternalSecretStores               bool `group:"Alpha Features:" help:"Enable support for External Secret Stores."`
 	EnableCompositionWebhookSchemaValidation bool `group:"Alpha Features:" help:"Enable support for Composition validation using schemas."`
 	EnableUsages                             bool `group:"Alpha Features:" help:"Enable support for deletion ordering and resource protection with Usages."`
+	EnableRealtimeCompositions               bool `group:"Alpha Features:" help:"Enable support for realtime compositions, i.e. watching MRs and reconciling compositions immediately when any MR is updated."`
```
[the remark this answers is not in the document]
Comment (negz) [reply cmd/crossplane/core/core.go:111]: While I'm nitpicking, "composed resources" would be more accurate than MRs. An XR can compose things like ProviderConfigs and other XRs, as well as MRs. I don't feel strongly about this one though - MRs seems like reasonable shorthand.

File: internal/controller/apiextensions/composite/reconciler.go [hunk 2]
```diff
@@ -593,6 +618,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 	}
 
+	if r.kindObserver != nil {
+		var gvks []schema.GroupVersionKind
+		for _, ref := range xr.GetResourceReferences() {
+			gvks = append(gvks, ref.GroupVersionKind())
+		}
+		r.kindObserver.Watch(gvks...)
```
Comment (negz) [inline internal/controller/apiextensions/composite/reconciler.go:626]: What if a GVK appears in the resource references but then is later removed? This could happen if the XR updates to a CompositionRevision that no longer produces a resource with that GVK. Do we need to stop the watch?

File: go.mod [hunk 3]
```diff
@@ -170,3 +170,5 @@ require (
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 )
+
+replace github.com/crossplane/crossplane-runtime => github.com/sttts/crossplane-runtime v0.0.0-20231006112446-53296455f75d
```
Comment (negz) [inline go.mod:174]: We can't merge this with this `replace` statement in place. 😄 Did the changes you need in c/crossplane-runtime get merged?

File: internal/controller/apiextensions/definition/indexes.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimeevent "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	compositeresource "github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
```
Comment (negz) [inline internal/controller/apiextensions/definition/indexes.go:33]: Would just `composite` work okay here? I don't think we alias this import anywhere else.

File: internal/controller/apiextensions/definition/indexes.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+// IndexCompositeResourceRefGVKs assumes the passed object is a composite. It
+// returns gvk keys for every resource referenced in the composite.
+func IndexCompositeResourceRefGVKs(o client.Object) []string {
+	u, ok := o.(*kunstructured.Unstructured)
+	if !ok {
+		return nil // should never happen
+	}
+	x := compositeresource.Unstructured{Unstructured: *u}
```
Comment (negz) [inline internal/controller/apiextensions/definition/indexes.go:53]:
```suggestion
	xr := compositeresource.Unstructured{Unstructured: *u}
```

I've been trying to slowly converge on `xr` as the variable name for composite resources (i.e. XRs) in our codebase, since it's pretty short and hopefully very obvious to folks.

File: internal/controller/apiextensions/definition/indexes.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	// from spec.resources. Hence, it will also work with functions.
+	compositeResourceRefGVKsIndex = "compositeResourceRefsGVKs"
+	// compositeResourcesRefsIndex is an index of resourceRefs that are owned
+	// by a composite.
+	compositeResourcesRefsIndex = "compositeResourcesRefs"
+)
+
+// IndexCompositeResourceRefGVKs assumes the passed object is a composite. It
+// returns gvk keys for every resource referenced in the composite.
+func IndexCompositeResourceRefGVKs(o client.Object) []string {
```
Comment (negz) [inline internal/controller/apiextensions/definition/indexes.go:48]: Is this and `IndexCompositeResourcesRefs` satisfying some interface? If so, is it worth testing that it does (mostly for documentation purposes)?

File: internal/controller/apiextensions/definition/managed.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+type managedResourceInformers struct {
```
Comment (negz) [inline internal/controller/apiextensions/definition/managed.go:44]: I'd suggest `s/managed/composed/g` for this whole thing. These informers are for composed resources. Managed resources are the most common set of composed resources, but not the only kind. These informers could potentially be handling XRs, ProviderConfigs, Usages, maybe other things in future.

File: internal/controller/apiextensions/definition/managed.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	log logging.Logger
+	mgr manager.Manager
+
+	mrInformersByCRDName map[schema.GroupVersionKind]mrInformer
```
Comment (negz) [inline internal/controller/apiextensions/definition/managed.go:48]: I assumed based on the whitespace around variables here that the two below maps were protected by the lock, but this one wasn't. That doesn't seem to be the case - maybe add a comment to `lock` explaining what it protects? Or just moving this down with the other maps would be enough of a hint for me personally.

Comment (negz) [reply internal/controller/apiextensions/definition/managed.go:48]: Also, is this actually `ByCRDName`? Personally I think `informers` would be fine - I can see in the type what the key is pretty easily.

File: internal/controller/apiextensions/definition/managed.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+type managedResourceInformers struct {
+	log logging.Logger
+	mgr manager.Manager
```
Comment (negz) [inline internal/controller/apiextensions/definition/managed.go:46]: Do we really need the whole manager in here? That's a pretty big interface if I recall correctly. Looks like you're only using `GetClient` and `GetConfig`.

File: internal/controller/apiextensions/definition/managed.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object)
+				new := newObj.(client.Object) //nolint:revive // this is fine
```
Comment (negz) [inline internal/controller/apiextensions/definition/managed.go:148]: Was the linter complaining about redefining `new`? If so, worth mentioning in the comment. (i.e. "Redefining new here is fine"). "This is fine" alone doesn't tell me much as a reader. 🙂

File: internal/controller/apiextensions/definition/managed.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+		// wait for in the background, and only when synced add to the routed cache
+		go func(gvk schema.GroupVersionKind) {
+			if synced := ca.WaitForCacheSync(ctx); synced {
+				log.Debug("Managed resource cache synced")
+				i.gvkRoutedCache.AddDelegate(gvk, ca)
+			}
+		}(gvk)
+	}
+} //nolint:govet // the context is stored and will be claused when the outer context is closed.
```
Comment (negz) [inline internal/controller/apiextensions/definition/managed.go:190]:
```suggestion
} //nolint:govet // the context is stored and will be closed when the outer context is closed.
```

🎅

**Review by negz — COMMENTED** [review]
I haven't fully reviewed yet, but I need to head out for the afternoon so here's what I have. Mostly nitpicks so far.

File: internal/controller/apiextensions/composite/reconciler.go (hunk 2 above)
Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:626]: There is another reconcile loop running every 10s removing old GVKs. Originally, I also started informer that, but changed that. So maybe we can reduce 10s to 30s or a minute or so.

File: go.mod (hunk 3 above)
Comment (sttts) [reply go.mod:174]: See https://github.com/crossplane/crossplane-runtime/pull/554. Not merged yet.

File: cmd/crossplane/core/core.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/crossplane/core/core.go:111]: Changed to "composed resources".

File: internal/controller/apiextensions/composite/reconciler.go (hunk 2 above)
Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:626]: Changed to 1m.

File: internal/controller/apiextensions/definition/indexes.go (hunk 4 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/indexes.go:33]: done

File: internal/controller/apiextensions/definition/indexes.go (hunk 5 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/indexes.go:53]: done

File: internal/controller/apiextensions/definition/indexes.go (hunk 6 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/indexes.go:48]: Not an interface, just a function type.

Comment (sttts) [reply internal/controller/apiextensions/definition/indexes.go:48]: Added a type assertion.

File: internal/controller/apiextensions/definition/managed.go (hunk 7 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/managed.go:44]: done

File: internal/controller/apiextensions/definition/managed.go (hunk 8 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/managed.go:48]: Agree, especially that they are indexed by GVK, not string. Done.

File: internal/controller/apiextensions/definition/managed.go (hunk 9 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/managed.go:46]: right, `Cluster` is enough.

File: internal/controller/apiextensions/definition/managed.go (hunk 10 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/managed.go:148]: renamed to `obj`.

File: internal/controller/apiextensions/definition/composed.go [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+	gvkRoutedCache *controller.GVKRoutedCache
+
+	lock      sync.RWMutex // everything below is protected by this lock
+	informers map[schema.GroupVersionKind]mrInformer
+	sinks     map[string]func(ev runtimeevent.UpdateEvent) // by some uid
+	xrs       map[string]xr                                // by controller name
+}
+
+type mrInformer struct {
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:56]:
```suggestion
type cdInformer struct {
```

I don't know that it's a very intuitive abbreviation, but FWIW we usually abbreviate `composed` to `cd` when we want shorter variable names.

File: internal/controller/apiextensions/definition/composed.go [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+type xr struct {
+	gvk   schema.GroupVersionKind
+	cache cache.Cache
+}
+
+var _ source.Source = &composedResourceInformers{}
+
+// Start implements source.Source, i.e. starting composedResourceInformers as
+// source with h as the sink of update events.
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:69]: Maybe mention that `Start` can be called more than once. This might be obvious to folks familiar with `source.Source`, but I was confused why `i.sinks` was a map that would seemingly only ever have a single entry until I realized `Start` could be called repeatedly to add new event handlers.

File: internal/controller/apiextensions/definition/composed.go [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+		i.lock.Lock()
+		defer i.lock.Unlock()
+		delete(i.sinks, id)
+	}()
+
+	return nil
+}
+
+// RegisterXR maps a controller name to a GVK.
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:95]:
```suggestion
// RegisterXRController maps a controller name to a GVK.
```

Perhaps `RegisterXRController` and `UnregisterXRController`? This would make it extra obvious to me that `name` is the name of the controller, not the `metadata.name` of an XR or something.

Comment (negz) [reply internal/controller/apiextensions/definition/composed.go:95]: Or actually, I guess it's really just an XR controller's cache and we happen to be using the controller name as a key. 🤔

File: internal/controller/apiextensions/definition/composed.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+	// start new informers
+	for _, gvk := range gvks {
+		if _, found := i.informers[gvk]; found {
+			continue
+		}
+
+		log := i.log.WithValues("gvk", gvk.String())
+
+		ctx, cancelFn := context.WithCancel(context.Background()) //nolint:govet // the outer context is closed by the caller
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:128]: Perhaps add a comment to remind future contributors to make sure they call `cancelFn()` in error cases? It could also be handy to explicitly point out that we `defer cancelFn` in the goroutine, which is why it's not being done here.

File: internal/controller/apiextensions/definition/composed.go [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+		// wait for in the background, and only when synced add to the routed cache
+		go func(gvk schema.GroupVersionKind) {
+			if synced := ca.WaitForCacheSync(ctx); synced {
+				log.Debug("Managed resource cache synced")
+				i.gvkRoutedCache.AddDelegate(gvk, ca)
+			}
+		}(gvk)
+	}
+} //nolint:govet // the context is stored and will be claused when the outer context is closed.
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:189]:
```suggestion
} //nolint:govet // the context is stored and will be closed when the outer context is closed.
```

File: internal/controller/apiextensions/composite/reconciler.go (hunk 2 above)
Comment (negz) [reply internal/controller/apiextensions/composite/reconciler.go:626]: Is it worth noting on the `Watch` interface that the implementation is responsible for garbage collecting watches that are no longer needed?

File: internal/controller/apiextensions/definition/composed.go [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+// UnregisterXR removes a controller name to GVK mapping.
+func (i *composedResourceInformers) UnregisterXR(name string) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	delete(i.xrs, name)
+}
+
+func (i *composedResourceInformers) Watch(gvks ...schema.GroupVersionKind) {
```
Comment (negz) [inline internal/controller/apiextensions/definition/composed.go:116]: A comment explaining what this method does would help. To replay my own understanding, I think:

* We create a new cache for each GVK we're told we should watch (e.g. each kind of composed resource)
* In practice that cache will be backed by a single informer, for said GVK.
* We get the informer for said GVK.
* We add an event handler to the informer that sends events to all registered sinks (i.e. event handlers)
* In practice each event handler will trigger a reconcile of any XR controller that composes this GVK.
* We tell the routed cache to delegate operations on this GVK to this cache.

File: internal/controller/apiextensions/definition/reconciler.go [hunk 18]
```diff
@@ -115,10 +124,18 @@ func (fn CRDRenderFn) Render(d *v1.CompositeResourceDefinition) (*extv1.CustomRe
...
-	r := NewReconciler(mgr,
+	r := NewReconciler(mgr, o,
 		WithLogger(o.Logger.WithValues("controller", name)),
-		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
-		WithOptions(o))
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	if o.Features.Enabled(features.EnableRealtimeCompositions) {
+		if err := mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
+			wait.UntilWithContext(ctx, r.xrInformers.reconcileManagedResourcesInformers, time.Minute)
```
Comment (negz) [inline internal/controller/apiextensions/definition/reconciler.go:133]: Maybe add a comment explaining that `reconcileManagedResourcesInformers` will run every minute until the context is cancelled. As someone who doesn't use `util/wait` very often I suspected that was the case but had to go look at its documentation to confirm. It's not super obvious from the function name/signature - "wait" doesn't sound like "run this repeatedly".

File: internal/controller/apiextensions/definition/managed.go (hunk 11 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/managed.go]: done

File: internal/controller/apiextensions/definition/indexes_test.go [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+							"name":       "xr",
+						},
+					},
+				},
+			},
+		}}, want: []string{"nop.crossplane.io/v1alpha1, Kind=NopResource", "nop.example.org/v1alpha1, Kind=NopResource"}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := IndexCompositeResourceRefGVKs(tt.args.object); !reflect.DeepEqual(got, tt.want) {
```
[the remark this answers is not in the document]
Comment (sttts) [reply internal/controller/apiextensions/definition/indexes_test.go:65]: done

File: internal/controller/apiextensions/definition/composed.go (hunk 12 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go:56]: done

File: internal/controller/apiextensions/definition/composed.go (hunk 13 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go:69]: done

File: internal/controller/apiextensions/definition/reconciler.go (hunk 18 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/reconciler.go:136]: done

File: internal/controller/apiextensions/definition/composed.go (hunk 17 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go]: done

**Comment by sttts** [comment]
@negz addressed all your comments, added tons of go docs, and rename things to be clearer. I like it.

File: internal/controller/apiextensions/definition/composed.go (hunk 15 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go:146]: done

File: go.mod (hunk 3 above)
Comment (negz) [reply go.mod]: FYI, just merged that.

**Comment by negz** [comment]
@sttts LGTM once the crossplane-runtime dependency is updated. Other than that, can you do a pass and `:s/cp/xr/g` for variables pertaining to composite resources. We're trying to move to the latter - I think folks find `xr == composite` to be more self documenting than `cp == composite`.

**Comment by sttts** [comment]
@negz did the cp->xr rename and removed the replace from go.mod.

File: test/e2e/realtimecompositions_test.go [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+				// 10 seconds is a long time for a realtime composition, but
+				// considerably below the normal reconcile time for a XR.
+				funcs.ResourcesHaveFieldValueWithin(10*time.Second, manifests, "claim.yaml", "status.coolerField", "I'M COOL!"),
+			).
+			WithTeardown("DeletePrerequisites", funcs.AllOf(
+				funcs.DeleteResources(manifests, "setup/*.yaml"),
+				funcs.ResourcesDeletedWithin(3*time.Minute, manifests, "setup/*.yaml"),
+			)).
+			// Disable our feature flag.
+			WithTeardown("DisableAlphaUsages", funcs.AllOf(
```
Comment (pedjak) [inline test/e2e/realtimecompositions_test.go:105]: you are not disabling usages feature here, right?

File: internal/controller/apiextensions/composite/reconciler.go [hunk 21]
```diff
@@ -606,6 +632,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 	}
 
+	if r.kindObserver != nil {
+		var gvks []schema.GroupVersionKind
+		for _, ref := range xr.GetResourceReferences() {
+			gvks = append(gvks, ref.GroupVersionKind())
+		}
+		r.kindObserver.WatchComposedResources(gvks...)
```
Comment (pedjak) [inline internal/controller/apiextensions/composite/reconciler.go:640]: if we now watch composed resources and trigger the reconcile in realtime, should we stop ending the loop with:

```golang
return reconcile.Result{RequeueAfter: r.pollInterval}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
```

and do instead:

```golang
return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)

```

Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:640]: We can't I guess. 1. environments 2. functions relying on 60s poll.

Should we do everything towards removing poll? Yes, please. @negz

File: test/e2e/realtimecompositions_test.go (hunk 20 above)
Comment (sttts) [reply test/e2e/realtimecompositions_test.go:105]: oops, great spot.

Comment (sttts) [reply test/e2e/realtimecompositions_test.go:105]: No, actually this was just cosmetics (feature name). Had no impact.

File: internal/controller/apiextensions/composite/reconciler.go (hunk 21 above)
Comment (pedjak) [reply internal/controller/apiextensions/composite/reconciler.go:640]: if environment feature is not enabled and composition is not based on functions, then we could? Also, we could watch environment for changes using the same `WatchComposeResource` approach, but keep in memory relationships between environments and compositions, to know later what composition to reconcile.

File: test/e2e/apiextensions_test.go [hunk 22]
```diff
@@ -53,6 +69,9 @@ func TestCompositionMinimal(t *testing.T) {
 			)).
 			Assess("CreateClaim", funcs.AllOf(
 				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.InBackground(funcs.LogResources(nopList)),
```
Comment (pedjak) [inline test/e2e/apiextensions_test.go:72]: how is this change related to the work done in this PR? I looks like a leftover of some debugging?

File: test/e2e/realtimecompositions_test.go [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+			)).
+			Assess("CreateClaim", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.InBackground(funcs.LogResources(nopList, withTestLabels)),
+				funcs.InBackground(funcs.LogResources(xnopList, withTestLabels)),
+				funcs.InBackground(funcs.LogResources(nopCrossplaneList, withTestLabels)),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+				funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available()),
+			)).
+			Assess("UpdateMR", funcs.AllOf(
```
Comment (pedjak) [inline test/e2e/realtimecompositions_test.go:85]: I think we can even avoid this step, because nop provider updates MRs even above, so line 83 would work even under a way shorter timeout.

File: internal/controller/apiextensions/composite/reconciler.go (hunk 21 above)
Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:640]: Am not against that direction, quite the opposite. Worth an attempt in a follow-up to see how far we get (in e2e).

For functions, I would like to see us being explicit that there is no 60s poll guarantee. Maybe we need an API to ask for one if necessary.

File: test/e2e/apiextensions_test.go (hunk 22 above)
Comment (sttts) [reply test/e2e/apiextensions_test.go:72]: These tests run under realtime composition as well. And in general, I would like us to have this kind of output to understand the core mechanisms at play in such a test. Hence, I added it also here. There are very likely more places where we need visibility of the main flow of a test.

File: test/e2e/realtimecompositions_test.go [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+			)).
+			WithSetup("PrerequisitesAreCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "setup/*.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "setup/*.yaml"),
+				funcs.ResourcesHaveConditionWithin(1*time.Minute, manifests, "setup/definition.yaml", apiextensionsv1.WatchingComposite()),
+				funcs.ResourcesHaveConditionWithin(2*time.Minute, manifests, "setup/provider.yaml", pkgv1.Healthy(), pkgv1.Active()),
+			)).
+			Assess("CreateClaim", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.InBackground(funcs.LogResources(nopList, withTestLabels)),
```
[the remark this answers is not in the document]
Comment (sttts) [reply test/e2e/realtimecompositions_test.go:79]: Not (only) for debugging. For visibility of the core mechanism at play in a test.

We cannot do these automatically as e.g. `ApplyResources` has no knowledge about what happens with those resources, i.e. the related resources they spawn.

File: test/e2e/realtimecompositions_test.go (hunk 23 above)
Comment (sttts) [reply test/e2e/realtimecompositions_test.go:85]: Don't think. I had to switch the nop manifest to immediate readiness, because it is unexpectedly slow to do so if you e.g. set it to `1s`. So now it is `0s`. And as a composition does reconciles frequently in the beginning for different reasons, I don't want to step into that noise. Hence, this step is explicitly changing something *after* readiness, when the noise is hopefully over.

File: test/e2e/realtimecompositions_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (pedjak) [reply test/e2e/realtimecompositions_test.go:79]: ApplyResource could become a bit smarter and understand that we are applying claim, and then via references follow to composite and MRs, or?

[the remark this answers is not in the document]
Comment (sttts) [reply test/e2e/realtimecompositions_test.go:79]: PR welcome for an automatic related object tracker :-)

File: test/e2e/realtimecompositions_test.go (hunk 20 above)
Comment (sttts) [reply test/e2e/realtimecompositions_test.go]: done

File: internal/controller/apiextensions/composite/reconciler.go (hunk 21 above)
Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:634]: That will be step 3: https://github.com/crossplane/crossplane/pull/4822

File: cmd/crossplane/core/core.go [hunk 25]
```diff
@@ -260,6 +261,10 @@ func (c *startCommand) Run(s *runtime.Scheme, log logging.Logger) error { //noli
 			TLSConfig: tcfg,
 		}
 	}
+	if c.EnableRealtimeCompositions {
+		o.Features.Enable(features.EnableRealtimeCompositions)
+		log.Info("Alpha feature enabled", "flag", features.EnableRealtimeCompositions)
+	}
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/crossplane/core/core.go:278]: https://github.com/crossplane/crossplane/issues/4828

**Review by phisco — COMMENTED** [review]
This should also delete the claim first and then check all nop resources have been deleted, no?
